### DISCUSSION TITLE
Regex support while packaging files with %ghost

### DISCUSSION THREAD

**Discussion by sshedi** [rpm-software-management/rpm#3480] [General]
Is it possible to ghost a regex pattern while packaging files?

A good use case of of this I can think of is, while packaging python modules we can ghost all byte compiled `.pyc` files and upon package removal, these `.pyc` files will get auto removed.

For instance, I want to do:
```
%files
...
%{python3_sitelib}/cloudinit/__pycache__/*.pyc
...
```

Fedora currently packages all `.pyc` files along with rpm which makes package size bigger and many byte compiled files might just be present in the system without ever getting used.

Second question in the same context is:
Is it possible to own a directory fully in rpm, upon removal; the directory should get removed without bothering about the content within.

For the same use case as above:
```
%files
...
%ghost %dir %{python3_sitelib}/cloudinit/__pycache__/
...
```
All the .pyc files get generated at runtime when used and if package gets removed, `__pycache__` dir should be gone entirely.

**Comment by sshedi** [comment]
I can patch each spec file and do this in `%postun` but it would be nice to have better solution for this problem.

**Comment by sshedi** [comment]
Hi @pmatilai, @ffesti any inputs on this?

**Comment by pmatilai** [comment]
> Is it possible to ghost a regex pattern while packaging files?

Did you actually try it?

> Is it possible to own a directory fully in rpm, upon removal; the directory should get removed without bothering about the content within.
> ...
> %ghost %dir %{python3_sitelib}/cloudinit/__pycache__/

Yes, it's very common. Just drop the %dir because that specifically tells rpm that this entry is ONLY about the directory node itself, and not the contents. Whether it actually works in combination with %ghost I have no idea.

As for your use-case for wanting to avoid bloating packages with redundant .pyc files: you might want to research the background as to why they are included in packages. There are multiple reasons for it, and none of them is "for fun".

**Comment by sshedi** [comment]
Sorry for the confusion.

> Did you actually try it?

I am aware that regex works while packaging files. But I want to use regex with ghost files.

```
%ghost %{python3_sitelib}/__pycache__
%ghost %{python3_sitelib}/__pycache__/*.pyc
```
Both `__pycache__` and the `*.pyc` are not present while the rpm is built and get created when I import the module after installing the rpm and I want those newly created pyc files and __pycache__ dir to go away when I remove the package by using ghost clause.

> Whether it actually works in combination with %ghost I have no idea.

This. I am trying it out and it doesn't work.

> As for your use-case for wanting to avoid bloating packages with redundant .pyc files: you might want to research the background as to why they are included in packages. There are multiple reasons for it, and none of them is "for fun".

I understand your point, it's just a standard we are following in our distro. Thanks for the inputs.

**Comment by pmatilai** [comment]
> This. I am trying it out and it doesn't work.

It's not clear to me what exactly you're trying and how it's not working (like, what happens?), because there are two different things getting conflated: regex on %ghost, and a whole directory with its contents (as ghost), and these are mutually exclusive.

**Comment by sshedi** [comment]
Here is the link to spec which I'm trying this with.

https://github.com/vmware/photon/blob/master/SPECS/selinux-python/selinux-python.spec#L52

I added below lines after line 52.
```
%ghost %{python3_sitelib}/__pycache__
%ghost %{python3_sitelib}/__pycache__/*.pyc
```

And after building and installing package, I did `python3 -c 'import selinux'` and it was successful.
Then I did `rpm -e selinux-python` and the pyc file is still left behind.

```
root@ph4dev [ ~ ]# ls /lib/python3.10/site-packages/selinux/__pycache__/
__init__.cpython-310.pyc
```
But if the ghost directives I used were effective, this should not have happened. Right?
